app/services/migration_service.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+class MigrationService
+  attr_reader :archive_path, :new_user_name
+  delegate :errors, :warnings, to: :archive_validator
+
+  def initialize(archive_path, new_user_name)
+    @archive_path = archive_path
+    @new_user_name = new_user_name
+  end
+
+  def validate
+    archive_validator.validate
+    raise ArchiveValidationFailed, errors.join("\n") if errors.any?
+    raise MigrationAlreadyExists if AccountMigration.where(old_person: old_person).any?
```

Comment:
You can't "delete" a user, once a users is created, it exists. You can close a user, but that can't be undone (so it can't be done again, at least not with the same new username).

So if the import was successful and the migration was created, everything was migrated away from the old user, and federated, it's migrated away from the old user on other pods too. So when you close your new user and want to import your old archive again, that won't work, since the old user doesn't own any posts anymore, everything was migrated to the new user. You would need to migrate from the new user to a third user if you aren't happy with your new user. And that's why this check is here.

Sure, you could close your new account, and create a new pod which did never receive the migration, and this check wouldn't fail there (since as you mentioned, it only checks locally), however, it doesn't mean that the migration works, it's basically a new users then, nothing is migrated on other pods, because other pods already migrated to the other new user. But there is no other check we can do here, that are the limitations of federated social networks :/

I don't know if I understood your questions correctly and if that answers your questions, so if not, please ask again :)

Reply:
> I don't know if I understood your questions correctly and if that answers your questions, so if not, please ask again :)

I think you got the nature of my question.  So the answer is we need to document that it's a one shot deal across the whole federation and we also need to make sure that if there is a failure part of the way through that it doesn't make it impossible to import the account a second time.  I'm thinking something like in the middle of an import the server crashes or something like that which would make all of their data tantalizingly close but really impossible to get to.

This will also make real world testing interesting since we have to continually hop accounts to recreate the process over and over again.

Reply:
> So the answer is we need to document that it's a one shot deal across the whole federation

The frontend needs a similar warning like the account closing (it can't be reverted, and in addition to migrations: it can only migrate to one target account). The rake task senya created here already has this warning.

> we also need to make sure that if there is a failure part of the way through that it doesn't make it impossible to import the account a second time.

First the validation runs (so that should already catch invalid archives), second the import runs (creating data from the archive) and then it triggers the migration locally (moving stuff from old user to new user), and as last step, it federates. So it should be pretty safe. You can always continue (even when you only moved half of the entities from the old user to the new), but you can't migrate to a second different users when you already started moving entities from the old user to the new user.

> I'm thinking something like in the middle of an import the server crashes or something like that which would make all of their data tantalizingly close but really impossible to get to.

I think the current state here should already support it, since it also retries the import if the target user already exists. We would need to also support that in the frontend to retry a failed import.

Also there is a second safety mechanism: If something really goes wrong and on some pods the migration fail, or whatever, the new user always works as it's own user (since it's a new key and new GUID), and the new account isn't corrupt then. Maybe it's missing some old data (it probably isn't complete anyway even with successful migration, since no pod would have had all past data and pods only migrate data they already know about the old user), but the future isn't broken. That was very important to me because that's one big problem when podmins try to rename stuff manually in the database and don't know about federation, then they usually break both, the old and the new user (forever), but with the migration here, the new users will always work for sure!

Reply:
> Maybe it's missing some old data (it probably isn't complete anyway even with successful migration, since no pod would have had all past data and pods only migrate data they already know about the old user), but the future isn't broken. 

I thought anything that was in the JSON data would be on the pod.  So let's say I'd move from some pod I'd been on for 5 years to my shiny new pod I just set up for myself.  I'm going to get everything that was on my old pod, all five years of data, not be limited to whatever the new pod had federated to it.  Is that not the case?

Reply:
> I'm going to get everything that was on my old pod, all five years of data, not be limited to whatever the new pod had federated to it.

That's the case, everything is imported to the pod you're migrating to, but all other pods only move the data they already know about your old user to your new user. And if on other pods something goes wrong (I don't know what should, but you never know, maybe some edge case with a very old post or whatever), then your new users isn't affected by that. Maybe not all data is migrated on other pods, but your new user just works anyway.

If something failed because of a bug that can be fixed (maybe needs an additional check or fix some inconsistent data from the past with a db migration), then the migration on other pods can still be retried (similar like we already retried account deletions that didn't run correctly because of a bug), but you don't need to wait until every pod retried your migration (if it's even fixable), your new user just works as a new user and you aren't blocked by that.

What I wanted to say that it's maybe incomplete on other pods anyway is: Your new pod gets everything from your JSON, but if your friend also creates his own pod, then that pod doesn't know anything about your old users, so nothing is migrated there. If your friends pod is already a month old, then the data from the last month is migrated.

Reply:
Okay whew.  I understand what you are saying.  From a user perspective that makes sense.